File: algorithmic-thinking/search-algorithms/dfs/dfs.c
/*
 * 深度优先搜索（DFS）实现 (C语言)
 * 
 * 特点：
 * - 优先往纵深方向探索
 * - 使用栈或递归
 * - 回溯机制
 * - 时间复杂度：O(V + E)
 * - 空间复杂度：O(V)
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

#define MAX_VERTICES 100
#define MAX_STACK_SIZE 100

// 栈结构
typedef struct {
    int items[MAX_STACK_SIZE];
    int top;
} Stack;

// 图结构（邻接表）
typedef struct Node {
    int vertex;
    struct Node* next;
} Node;

typedef struct {
    Node* adjLists[MAX_VERTICES];
    int numVertices;
} Graph;

// 初始化栈
void initStack(Stack* s) {
    s->top = -1;
}

// 检查栈是否为空
bool isStackEmpty(Stack* s) {
    return s->top == -1;
}

// 入栈
void push(Stack* s, int value) {
    if (s->top == MAX_STACK_SIZE - 1) {
        printf("栈已满\n");
        return;
    }
    s->top++;
    s->items[s->top] = value;
}

// 出栈
int pop(Stack* s) {
    if (isStackEmpty(s)) {
        printf("栈为空\n");
        return -1;
    }
    int item = s->items[s->top];
    s->top--;
    return item;
}

// 创建新节点
Node* createNode(int v) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->vertex = v;
    newNode->next = NULL;
    return newNode;
}

// 初始化图
void initGraph(Graph* graph, int vertices) {
    graph->numVertices = vertices;
    
    for (int i = 0; i < vertices; i++) {
        graph->adjLists[i] = NULL;
    }
}

// 添加边（无向图）
void addEdge(Graph* graph, int src, int dest) {
    // 添加 src -> dest 的边
    Node* newNode = createNode(dest);
    newNode->next = graph->adjLists[src];
    graph->adjLists[src] = newNode;
    
    // 添加 dest -> src 的边（无向图）
    newNode = createNode(src);
    newNode->next = graph->adjLists[dest];
    graph->adjLists[dest] = newNode;
}

// DFS递归实现
void dfsRecursive(Graph* graph, int vertex, bool visited[]) {
    // 标记当前顶点为已访问
    visited[vertex] = true;
    printf("%d ", vertex);
    
    // 递归访问所有未访问的邻接顶点
    Node* temp = graph->adjLists[vertex];
    while (temp != NULL) {
        int adjVertex = temp->vertex;
        
        if (!visited[adjVertex]) {
            dfsRecursive(graph, adjVertex, visited);
        }
        
        temp = temp->next;
    }
}

// DFS递归遍历（包装函数）
void dfsRecursiveTraversal(Graph* graph, int startVertex) {
    bool visited[MAX_VERTICES] = {false};
    
    printf("DFS递归遍历结果: ");
    dfsRecursive(graph, startVertex, visited);
    printf("\n");
}

// DFS迭代实现（使用栈）
void dfsIterative(Graph* graph, int startVertex) {
    bool visited[MAX_VERTICES] = {false};
    Stack stack;
    
    initStack(&stack);
    
    // 将起始顶点入栈
    push(&stack, startVertex);
    
    printf("DFS迭代遍历结果: ");
    
    while (!isStackEmpty(&stack)) {
        // 出栈一个顶点
        int currentVertex = pop(&stack);
        
        // 如果该顶点未被访问
        if (!visited[currentVertex]) {
            visited[currentVertex] = true;
            printf("%d ", currentVertex);
            
            // 将所有未访问的邻接顶点入栈
            // 注意：为了保持与递归相似的顺序，需要反向入栈
            Node* temp = graph->adjLists[currentVertex];
            
            // 先收集所有邻接顶点
            int neighbors[MAX_VERTICES];
            int neighborCount = 0;
            
            while (temp != NULL) {
                int adjVertex = temp->vertex;
                if (!visited[adjVertex]) {
                    neighbors[neighborCount++] = adjVertex;
                }
                temp = temp->next;
            }
            
            // 反向入栈以保持顺序
            for (int i = neighborCount - 1; i >= 0; i--) {
                push(&stack, neighbors[i]);
            }
        }
    }
    
    printf("\n");
}

// DFS查找路径
bool dfsFindPath(Graph* graph, int current, int target, bool visited[], int path[], int* pathIndex) {
    // 标记当前顶点为已访问
    visited[current] = true;
    path[*pathIndex] = current;
    (*pathIndex)++;
    
    // 如果找到目标
    if (current == target) {
        return true;
    }
    
    // 递归访问所有邻接顶点
    Node* temp = graph->adjLists[current];
    while (temp != NULL) {
        int adjVertex = temp->vertex;
        
        if (!visited[adjVertex]) {
            if (dfsFindPath(graph, adjVertex, target, visited, path, pathIndex)) {
                return true;
            }
        }
        
        temp = temp->next;
    }
    
    // 回溯：从路径中移除当前顶点
    (*pathIndex)--;
    return false;
}

// 查找两点之间的路径
void findPath(Graph* graph, int start, int end) {
    bool visited[MAX_VERTICES] = {false};
    int path[MAX_VERTICES];
    int pathIndex = 0;
    
    printf("从 %d 到 %d 的路径查找:\n", start, end);
    
    if (dfsFindPath(graph, start, end, visited, path, &pathIndex)) {
        printf("找到路径: ");
        for (int i = 0; i < pathIndex; i++) {
            printf("%d", path[i]);
            if (i < pathIndex - 1) printf(" -> ");
        }
        printf("\n");
    } else {
        printf("未找到路径\n");
    }
}

// DFS检测环
bool dfsDetectCycle(Graph* graph, int vertex, int parent, bool visited[]) {
    visited[vertex] = true;
    
    Node* temp = graph->adjLists[vertex];
    while (temp != NULL) {
        int adjVertex = temp->vertex;
        
        // 如果邻接顶点未被访问，递归检查
        if (!visited[adjVertex]) {
            if (dfsDetectCycle(graph, adjVertex, vertex, visited)) {
                return true;
            }
        }
        // 如果邻接顶点已被访问且不是父节点，则存在环
        else if (adjVertex != parent) {
            return true;
        }
        
        temp = temp->next;
    }
    
    return false;
}

// 检测图中是否存在环
bool hasCycle(Graph* graph) {
    bool visited[MAX_VERTICES] = {false};
    
    for (int i = 0; i < graph->numVertices; i++) {
        if (!visited[i]) {
            if (dfsDetectCycle(graph, i, -1, visited)) {
                return true;
            }
        }
    }
    
    return false;
}

// DFS计算连通分量
void dfsConnectedComponents(Graph* graph) {
    bool visited[MAX_VERTICES] = {false};
    int componentCount = 0;
    
    printf("连通分量:\n");
    
    for (int i = 0; i < graph->numVertices; i++) {
        if (!visited[i]) {
            componentCount++;
            printf("分量 %d: ", componentCount);
            
            // 使用DFS遍历当前连通分量
            dfsRecursive(graph, i, visited);
            printf("\n");
        }
    }
    
    printf("总连通分量数: %d\n", componentCount);
}

// 释放图内存
void freeGraph(Graph* graph) {
    for (int i = 0; i < graph->numVertices; i++) {
        Node* temp = graph->adjLists[i];
        while (temp != NULL) {
            Node* toDelete = temp;
            temp = temp->next;
            free(toDelete);
        }
    }
}

int main() {
    printf("=== 深度优先搜索（DFS）演示 ===\n\n");
    
    // 创建图
    Graph graph;
    int vertices = 8;
    initGraph(&graph, vertices);
    
    // 添加边
    addEdge(&graph, 0, 1);
    addEdge(&graph, 0, 2);
    addEdge(&graph, 1, 3);
    addEdge(&graph, 1, 4);
    addEdge(&graph, 2, 5);
    addEdge(&graph, 2, 6);
    addEdge(&graph, 3, 7);
    addEdge(&graph, 4, 7);
    addEdge(&graph, 5, 6);
    
    printf("图结构:\n");
    printf("0 -- 1 -- 3 -- 7\n");
    printf("|    |         |\n");
    printf("|    |         |\n");
    printf("2 -- 4 --------|\n");
    printf("|    |\n");
    printf("|    |\n");
    printf("5 -- 6\n\n");
    
    // 1. DFS递归遍历
    printf("1. DFS递归遍历（从顶点0开始）\n");
    dfsRecursiveTraversal(&graph, 0);
    printf("\n");
    
    // 2. DFS迭代遍历
    printf("2. DFS迭代遍历（从顶点0开始）\n");
    dfsIterative(&graph, 0);
    printf("\n");
    
    // 3. 查找路径
    printf("3. 查找路径\n");
    findPath(&graph, 0, 7);
    printf("\n");
    
    // 4. 检测环
    printf("4. 检测环\n");
    if (hasCycle(&graph)) {
        printf("图中存在环\n");
    } else {
        printf("图中不存在环\n");
    }
    printf("\n");
    
    // 5. 计算连通分量
    printf("5. 计算连通分量\n");
    dfsConnectedComponents(&graph);
    printf("\n");
    
    // 6. 从不同起点开始DFS
    printf("6. 从不同起点开始DFS\n");
    printf("从顶点3开始（递归）: ");
    bool visited[MAX_VERTICES] = {false};
    dfsRecursive(&graph, 3, visited);
    printf("\n");
    
    // 释放内存
    freeGraph(&graph);
    
    return 0;
}

File: algorithmic-thinking/search-algorithms/dfs/dfs.rs
// 深度优先搜索（DFS）实现 (Rust)
//
// 特点：
// - 优先往纵深方向探索
// - 使用栈或递归
// - 回溯机制
// - 时间复杂度：O(V + E)
// - 空间复杂度：O(V)

use std::collections::VecDeque;

// 图结构（邻接表）
#[derive(Debug, Clone)]
pub struct Graph {
    adj_lists: Vec<Vec<usize>>,
    num_vertices: usize,
}

impl Graph {
    // 初始化图
    pub fn new(vertices: usize) -> Self {
        Graph {
            adj_lists: vec![Vec::new(); vertices],
            num_vertices: vertices,
        }
    }

    // 添加边（无向图）
    pub fn add_edge(&mut self, src: usize, dest: usize) {
        self.adj_lists[src].push(dest);
        self.adj_lists[dest].push(src); // 无向图
    }

    // DFS递归实现
    fn dfs_recursive(&self, vertex: usize, visited: &mut [bool], result: &mut Vec<usize>) {
        // 标记当前顶点为已访问
        visited[vertex] = true;
        result.push(vertex);

        // 递归访问所有未访问的邻接顶点
        for &adj_vertex in &self.adj_lists[vertex] {
            if !visited[adj_vertex] {
                self.dfs_recursive(adj_vertex, visited, result);
            }
        }
    }

    // DFS递归遍历（包装函数）
    pub fn dfs_recursive_traversal(&self, start_vertex: usize) -> Vec<usize> {
        let mut visited = vec![false; self.num_vertices];
        let mut result = Vec::new();
        
        self.dfs_recursive(start_vertex, &mut visited, &mut result);
        result
    }

    // DFS迭代实现（使用栈）
    pub fn dfs_iterative(&self, start_vertex: usize) -> Vec<usize> {
        let mut visited = vec![false; self.num_vertices];
        let mut stack = Vec::new();
        let mut result = Vec::new();

        // 将起始顶点入栈
        stack.push(start_vertex);

        while let Some(current_vertex) = stack.pop() {
            // 如果该顶点未被访问
            if !visited[current_vertex] {
                visited[current_vertex] = true;
                result.push(current_vertex);

                // 将所有未访问的邻接顶点入栈
                // 注意：为了保持与递归相似的顺序，需要反向入栈
                for &adj_vertex in self.adj_lists[current_vertex].iter().rev() {
                    if !visited[adj_vertex] {
                        stack.push(adj_vertex);
                    }
                }
            }
        }

        result
    }

    // DFS查找路径
    fn dfs_find_path(
        &self,
        current: usize,
        target: usize,
        visited: &mut [bool],
        path: &mut Vec<usize>,
    ) -> bool {
        // 标记当前顶点为已访问
        visited[current] = true;
        path.push(current);

        // 如果找到目标
        if current == target {
            return true;
        }

        // 递归访问所有邻接顶点
        for &adj_vertex in &self.adj_lists[current] {
            if !visited[adj_vertex] {
                if self.dfs_find_path(adj_vertex, target, visited, path) {
                    return true;
                }
            }
        }

        // 回溯：从路径中移除当前顶点
        path.pop();
        false
    }

    // 查找两点之间的路径
    pub fn find_path(&self, start: usize, end: usize) -> Option<Vec<usize>> {
        let mut visited = vec![false; self.num_vertices];
        let mut path = Vec::new();

        if self.dfs_find_path(start, end, &mut visited, &mut path) {
            Some(path)
        } else {
            None
        }
    }

    // DFS检测环
    fn dfs_detect_cycle(&self, vertex: usize, parent: usize, visited: &mut [bool]) -> bool {
        visited[vertex] = true;

        for &adj_vertex in &self.adj_lists[vertex] {
            // 如果邻接顶点未被访问，递归检查
            if !visited[adj_vertex] {
                if self.dfs_detect_cycle(adj_vertex, vertex, visited) {
                    return true;
                }
            }
            // 如果邻接顶点已被访问且不是父节点，则存在环
            else if adj_vertex != parent {
                return true;
            }
        }

        false
    }

    // 检测图中是否存在环
    pub fn has_cycle(&self) -> bool {
        let mut visited = vec![false; self.num_vertices];

        for i in 0..self.num_vertices {
            if !visited[i] {
                if self.dfs_detect_cycle(i, self.num_vertices, &mut visited) {
                    return true;
                }
            }
        }

        false
    }

    // DFS计算连通分量
    pub fn dfs_connected_components(&self) -> Vec<Vec<usize>> {
        let mut visited = vec![false; self.num_vertices];
        let mut components = Vec::new();

        for i in 0..self.num_vertices {
            if !visited[i] {
                let mut component = Vec::new();
                self.dfs_recursive(i, &mut visited, &mut component);
                components.push(component);
            }
        }

        components
    }

    // DFS拓扑排序（用于有向无环图）
    pub fn dfs_topological_sort(&self) -> Vec<usize> {
        let mut visited = vec![false; self.num_vertices];
        let mut stack = Vec::new();

        // 对每个未访问的顶点进行DFS
        for i in 0..self.num_vertices {
            if !visited[i] {
                self.topological_sort_util(i, &mut visited, &mut stack);
            }
        }

        // 弹出栈中的元素得到拓扑排序
        stack.reverse();
        stack
    }

    // 拓扑排序的辅助函数
    fn topological_sort_util(&self, vertex: usize, visited: &mut [bool], stack: &mut Vec<usize>) {
        visited[vertex] = true;

        for &adj_vertex in &self.adj_lists[vertex] {
            if !visited[adj_vertex] {
                self.topological_sort_util(adj_vertex, visited, stack);
            }
        }

        // 将顶点入栈
        stack.push(vertex);
    }

    // 获取转置图（用于强连通分量）
    fn get_transpose(&self) -> Graph {
        let mut transposed = Graph::new(self.num_vertices);

        for v in 0..self.num_vertices {
            for &adj_vertex in &self.adj_lists[v] {
                transposed.adj_lists[adj_vertex].push(v);
            }
        }

        transposed
    }

    // 填充顺序的辅助函数
    fn fill_order(&self, vertex: usize, visited: &mut [bool], stack: &mut Vec<usize>) {
        visited[vertex] = true;

        for &adj_vertex in &self.adj_lists[vertex] {
            if !visited[adj_vertex] {
                self.fill_order(adj_vertex, visited, stack);
            }
        }

        stack.push(vertex);
    }

    // DFS检测强连通分量（用于有向图）
    pub fn dfs_strongly_connected_components(&self) -> Vec<Vec<usize>> {
        // Kosaraju算法
        let mut visited = vec![false; self.num_vertices];
        let mut stack = Vec::new();

        // 第一步：对原图进行DFS，将顶点按完成时间入栈
        for i in 0..self.num_vertices {
            if !visited[i] {
                self.fill_order(i, &mut visited, &mut stack);
            }
        }

        // 第二步：创建转置图
        let transposed = self.get_transpose();

        // 第三步：按栈中顺序对转置图进行DFS
        visited = vec![false; self.num_vertices];
        let mut components = Vec::new();

        while let Some(vertex) = stack.pop() {
            if !visited[vertex] {
                let mut component = Vec::new();
                transposed.dfs_recursive(vertex, &mut visited, &mut component);
                components.push(component);
            }
        }

        components
    }

    // 检查是否为二分图
    pub fn is_bipartite(&self) -> bool {
        let mut color = vec![0; self.num_vertices]; // 0: 未染色, 1: 染色A, -1: 染色B

        for i in 0..self.num_vertices {
            if color[i] == 0 {
                let mut stack = Vec::new();
                color[i] = 1;
                stack.push(i);

                while let Some(current_vertex) = stack.pop() {
                    for &adj_vertex in &self.adj_lists[current_vertex] {
                        if color[adj_vertex] == 0 {
                            color[adj_vertex] = -color[current_vertex];
                            stack.push(adj_vertex);
                        } else if color[adj_vertex] == color[current_vertex] {
                            return false;
                        }
                    }
                }
            }
        }

        true
    }
}

fn main() {
    println!("=== 深度优先搜索（DFS）演示 ===\n");

    // 创建图
    let mut graph = Graph::new(8);

    // 添加边
    graph.add_edge(0, 1);
    graph.add_edge(0, 2);
    graph.add_edge(1, 3);
    graph.add_edge(1, 4);
    graph.add_edge(2, 5);
    graph.add_edge(2, 6);
    graph.add_edge(3, 7);
    graph.add_edge(4, 7);
    graph.add_edge(5, 6);

    println!("图结构:");
    println!("0 -- 1 -- 3 -- 7");
    println!("|    |         |");
    println!("|    |         |");
    println!("2 -- 4 --------|");
    println!("|    |");
    println!("|    |");
    println!("5 -- 6\n");

    // 1. DFS递归遍历
    println!("1. DFS递归遍历（从顶点0开始）");
    let recursive_result = graph.dfs_recursive_traversal(0);
    println!("结果: {:?}", recursive_result);
    println!();

    // 2. DFS迭代遍历
    println!("2. DFS迭代遍历（从顶点0开始）");
    let iterative_result = graph.dfs_iterative(0);
    println!("结果: {:?}", iterative_result);
    println!();

    // 3. 查找路径
    println!("3. 查找路径");
    match graph.find_path(0, 7) {
        Some(path) => println!("从 0 到 7 的路径: {:?}", path),
        None => println!("未找到从 0 到 7 的路径"),
    }
    println!();

    // 4. 检测环
    println!("4. 检测环");
    if graph.has_cycle() {
        println!("图中存在环");
    } else {
        println!("图中不存在环");
    }
    println!();

    // 5. 计算连通分量
    println!("5. 计算连通分量");
    let components = graph.dfs_connected_components();
    println!("连通分量数: {}", components.len());
    for (i, component) in components.iter().enumerate() {
        println!("分量 {}: {:?}", i + 1, component);
    }
    println!();

    // 6. 从不同起点开始DFS
    println!("6. 从不同起点开始DFS");
    println!("从顶点3开始（递归）: {:?}", graph.dfs_recursive_traversal(3));
    println!("从顶点3开始（迭代）: {:?}", graph.dfs_iterative(3));

    // 7. 检查是否为二分图
    println!("\n7. 检查是否为二分图");
    if graph.is_bipartite() {
        println!("该图是二分图");
    } else {
        println!("该图不是二分图");
    }
}
